Publish C++ developer docs (#11475)
rapidsai/cudf: cpp/doxygen/developer_guide/DOCUMENTATION.md

### @brief

> The [\@brief](http://www.doxygen.nl/manual/commands.html#cmdbrief) text should be a short, one sentence description.

Reviewer: ```suggestion
The [\@brief](https://www.doxygen.nl/manual/commands.html#cmdbrief) text should be a short, one sentence description.
```

Doxygen does not provide much space to show this text in the output pages.
Always follow the \@brief line with a blank comment line.

The longer description is the rest of the comment text that is not tagged with any doxygen command.

    /**
     * @brief Short description.
     *
     * Long description.
     *

### \@copydoc

Documentation for declarations in headers should be clear and complete.

> You can use the [\@copydoc](http://www.doxygen.nl/manual/commands.html#cmdcopydoc) tag to avoid duplicating the comment block for a function definition.

Reviewer: ```suggestion
You can use the [\@copydoc](https://www.doxygen.nl/manual/commands.html#cmdcopydoc) tag to avoid duplicating the comment block for a function definition.
```

| Command | Description |
| ------- | ----------- |
| [\@throw](#throw) | Specify the conditions in which the function may throw an exception |
| [\@tparam](#tparam) | Description for each template parameter |
| [\@param](#param) | Description for each function parameter |
| [\@return](#return) | Short description of object or value returned |

#### \@throw

> Add an [\@throw](http://www.doxygen.nl/manual/commands.html#cmdthrow) comment line in the doxygen block for each exception that the function may throw.

Reviewer: ```suggestion
Add an [\@throw](https://www.doxygen.nl/manual/commands.html#cmdthrow) comment line in the doxygen block for each exception that the function may throw.
```

You only need to include exceptions thrown by the function itself.
If the function calls another function that may throw an exception, you do not need to document those exceptions here.

Include the name of the exception without backtick marks so doxygen can add reference links correctly.

     *
     * @throw cudf::logic_error if `input_argument` is negative or zero
     *

Using \@throws is also acceptable but VS Code and other tools only do syntax highlighting on \@throw.

#### @tparam

> Add a [\@tparam](http://www.doxygen.nl/manual/commands.html#cmdtparam) comment line for each template parameter declared by this function.

Reviewer: Use `https:` everywhere.

Doxygen supports syntax highlighting for C++ and several other programming languages (e.g. Python, Java).
By default, the \@code tag uses syntax highlighting based on the source code in which it is found.

     *
     * @code
     * auto result = cudf::make_column( );
     * @endcode
     *

You can specify a different language by indicating the file extension in the tag:

     *
     * @code{.py}
     * import cudf
     * s = cudf.Series([1,2,3])
     * @endcode
     *

If you wish to use pseudocode in your example, use the following:

     *
     * Sometimes pseudocode is clearer.
     * @code{.pseudo}
     * s = int column of [ 1, 2, null, 4 ]
     * r = fill( s, [1, 2], 0 )
     * r is now [ 1, 0, 0, 4 ]
     * @endcode
     *

When writing example snippets, using fully qualified class names allows doxygen to add reference links to the example.

     *
     * @code
     * auto result1 = make_column( ); // reference link will not be created
     * auto result2 = cudf::make_column( ); // reference link will be created
     * @endcode
     *

> Although using 3 backtick marks \`\`\` for example blocks will work too, they do not standout as well in VS Code and other source editors.

Reviewer: ```suggestion
Although using 3 backtick marks \`\`\` for example blocks will work too, they do not stand out as well in VS Code and other source editors.
```

Modify this file only to add or update groups.
The existing groups have been carefully structured and named, so new groups should be added thoughtfully.

When creating a new API, specify its group using the [\@ingroup](http://www.doxygen.nl/manual/commands.html#cmdingroup) tag and the group reference id from the [doxygen_groups.h](../include/doxygen_groups.h) file.

    namespace cudf {

    /**
     * @brief ...
     *
     * @ingroup transformation_fill
     *
     * @param ...
     * @return ...
     */
    std::unique_ptr<column> fill(table_view const& input,...);

    }  // namespace cudf

> You can also use the \@addtogroup with a `@{ ... @}` pair to automatically include  doxygen comment blocks as part of a group.

Reviewer: ```suggestion
You can also use the \@addtogroup with a `@{ ... @}` pair to automatically include doxygen comment blocks as part of a group.
```
